iothub/device/src/ModuleClient.cs

```diff
@@ -0,0 +1,511 @@
+    /// <param name="userContext">The context object passed in</param>
+    /// <returns>MessageResponse</returns>
+    public delegate Task<MessageResponse> MessageHandler(Message message, object userContext);
+
+    /// <summary>
+    /// Contains methods that a device can use to send messages to and receive from the service.
+    /// </summary>
+    public sealed class ModuleClient : BaseClient
+    {
+        volatile Dictionary<string, Tuple<MessageHandler, object>> receiveEventEndpoints;
+
+        readonly SemaphoreSlim receiveSemaphore = new SemaphoreSlim(1, 1);
+
+        volatile Tuple<MessageHandler, object> defaultEventCallback;
+
+        private ProductInfo productInfo = new ProductInfo();
+
+        /// <summary>
+        /// Stores custom product information that will be appended to the user agent string that is sent to IoT Hub.
+        /// </summary>
+        public string ProductInfo
+        {
+            // We store DeviceClient.ProductInfo as a string property of an object (rather than directly as a string)
+            // so that updates will propagate down to the transport layer throughout the lifetime of the DeviceClient
+            // object instance.
+            get => productInfo.Extra;
+            set => productInfo.Extra = value;
+        }
+
+        ModuleClient(IotHubConnectionString iotHubConnectionString, ITransportSettings[] transportSettings, IDeviceClientPipelineBuilder pipelineBuilder)
+            : base(iotHubConnectionString, transportSettings)
+        {
+            if (string.IsNullOrWhiteSpace(iotHubConnectionString.ModuleId))
+            {
+                throw new ArgumentException("ModuleId not present - ModuleClient can be used only with modules");
+            }
+
+            var pipelineContext = new PipelineContext();
+            pipelineContext.Set(transportSettings);
+            pipelineContext.Set(iotHubConnectionString);
+            pipelineContext.Set<OnMethodCalledDelegate>(OnMethodCalled);
+            pipelineContext.Set<Action<TwinCollection>>(OnReportedStatePatchReceived);
+            pipelineContext.Set<OnConnectionClosedDelegate>(OnConnectionClosed);
+            pipelineContext.Set<OnConnectionOpenedDelegate>(OnConnectionOpened);
+            pipelineContext.Set<OnReceiveEventMessageCalledDelegate>(OnReceiveEventMessageCalled);
+            pipelineContext.Set(this.productInfo);
+
+            IDelegatingHandler innerHandler = pipelineBuilder.Build(pipelineContext);
+            this.InnerHandler = innerHandler;
+        }
+
+        /// <summary>
+        /// Create an Amqp ModuleClient from individual parameters
+        /// </summary>
+        /// <param name="hostname">The fully-qualified DNS hostname of IoT Hub</param>
+        /// <param name="authenticationMethod">The authentication method that is used</param>
+        /// <returns>ModuleClient</returns>
+        public static ModuleClient Create(string hostname, IAuthenticationMethod authenticationMethod)
```

Comment:
Can we pull in all the common code for the Create...() functions into the Base Class as well?

Reply:
I couldn't because it finally calls ModuleClient.Create or DeviceClient.Create, and pulling it into Base would require exposing some methods. 
But I am planning a further refactor of this code... will have that iteration out by tonight.

Reply:
Refactored these into a separate ClientFactory class.